PyHeapTypeObject errors with PyPy

(Originally sent to swig-user mailing list)

Our conda-forge feedstock received an automatic PR to add PyPy support. Unfortunately, I get a lot of errors when compiling the SWIG wrapper like this:
```
$SRC_DIR/build/src/swig/CMakeFiles/_stir.dir/stirPYTHON_wrap.cxx:158287:1:
   error: too many initializers for 'PyHeapTypeObject' {aka '_heaptypeobject'}
```
This is with swig 4.0.2 and pypy 3.8. You can find the log at [Pipelines - Run 20220414.1 logs (azure.com)](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=492070&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=b2a8456a-fb11-5506-ca32-5ccd32538dc0&l=6141)

I find the following in the wrapped file (EDIT: generated locally by swig 3.0.12)
```
#if PY_VERSION_HEX < 0x02030000
typedef struct {
  PyTypeObject type;
  PyNumberMethods as_number;
  PyMappingMethods as_mapping;
  PySequenceMethods as_sequence;
  PyBufferProcs as_buffer;
  PyObject *name, *slots;
} PyHeapTypeObject;
#endif

/// stuff

static PyHeapTypeObject SwigPyBuiltin__swig__SwigPyIterator_type = {
  {
#if PY_VERSION_HEX >= 0x03000000
    PyVarObject_HEAD_INIT(NULL, 0)
#else
    PyObject_HEAD_INIT(NULL)
    0,                                        /* ob_size */
#endif
    "stir.SwigPyIterator",                    /* tp_name */
    sizeof(SwigPyObject),                     /* tp_basicsize */
    0,                                        /* tp_itemsize */
    _wrap_delete_SwigPyIterator_destructor_closure,               /* tp_dealloc */
    (printfunc) 0,                            /* tp_print */
    (getattrfunc) 0,                          /* tp_getattr */
    /// more stuff
```

I cannot find any reference to `PyHeapTypeObject` but wonder if the `#if PY_VERSION_HEX` statements are failing with PyPy.

I've now installed swig 4.0.2 via conda on Windows to generate the wrapper. Compilation with VS 2019 gives a different error, but related to the same code (edited to remove the file path for brevity):
```
1>...\stirPYTHON_wrap.cxx(156696,17): error C2440: 'initializing': cannot convert from 'printfunc' to 'Py_ssize_t'
1>...\stirPYTHON_wrap.cxx(156696,17): message : There is no context in which this conversion is possible
```
This points to the middle line in generated code like
```
    _wrap_delete_ios_destructor_closure,      
    (printfunc) 0,                            
    (getattrfunc) 0,                          
```

I've also manually applied the patch from #1843 to my wrapped file, but that didn't help.

--- a/Lib/python/builtin.swg
+++ b/Lib/python/builtin.swg
@@ -211,7 +211,11 @@ SwigPyStaticVar_Type(void) {
       sizeof(PyGetSetDescrObject),              /* tp_basicsize */
       0,                                        /* tp_itemsize */
       (destructor)SwigPyStaticVar_dealloc,      /* tp_dealloc */
-      0,                                        /* tp_print */
+#if PY_VERSION_HEX < 0x030800b4
+      (printfunc)0,                             /*tp_print*/
+#else
+      (Py_ssize_t)0,                            /*tp_vectorcall_offset*/
+#endif
       0,                                        /* tp_getattr */
       0,                                        /* tp_setattr */
       0,                                        /* tp_compare */
@@ -295,7 +299,11 @@ SwigPyObjectType(void) {
       PyType_Type.tp_basicsize,                 /* tp_basicsize */
       0,                                        /* tp_itemsize */
       0,                                        /* tp_dealloc */
-      0,                                        /* tp_print */
+#if PY_VERSION_HEX < 0x030800b4
+      (printfunc)0,                             /*tp_print*/
+#else
+      (Py_ssize_t)0,                            /*tp_vectorcall_offset*/
+#endif
       0,                                        /* tp_getattr */
       0,                                        /* tp_setattr */
       0,                                        /* tp_compare */

--- a/Lib/python/builtin.swg
+++ b/Lib/python/builtin.swg
@@ -212,9 +212,9 @@ SwigPyStaticVar_Type(void) {
       0,                                        /* tp_itemsize */
       (destructor)SwigPyStaticVar_dealloc,      /* tp_dealloc */
 #if PY_VERSION_HEX < 0x030800b4
-      (printfunc)0,                             /*tp_print*/
+      (printfunc)0,                             /* tp_print */
 #else
-      (Py_ssize_t)0,                            /*tp_vectorcall_offset*/
+      (Py_ssize_t)0,                            /* tp_vectorcall_offset */
 #endif
       0,                                        /* tp_getattr */
       0,                                        /* tp_setattr */
@@ -300,9 +300,9 @@ SwigPyObjectType(void) {
       0,                                        /* tp_itemsize */
       0,                                        /* tp_dealloc */
 #if PY_VERSION_HEX < 0x030800b4
-      (printfunc)0,                             /*tp_print*/
+      (printfunc)0,                             /* tp_print */
 #else
-      (Py_ssize_t)0,                            /*tp_vectorcall_offset*/
+      (Py_ssize_t)0,                            /* tp_vectorcall_offset */
 #endif
       0,                                        /* tp_getattr */
       0,                                        /* tp_setattr */

--- a/Lib/python/pyrun.swg
+++ b/Lib/python/pyrun.swg
@@ -348,7 +348,11 @@ swig_varlink_type(void) {
       sizeof(swig_varlinkobject),         /* tp_basicsize */
       0,                                  /* tp_itemsize */
       (destructor) swig_varlink_dealloc,  /* tp_dealloc */
-      0,                                  /* tp_print */
+#if PY_VERSION_HEX < 0x030800b4
+      (printfunc)0,                       /*tp_print*/
+#else
+      (Py_ssize_t)0,                      /*tp_vectorcall_offset*/
+#endif
       (getattrfunc) swig_varlink_getattr, /* tp_getattr */
       (setattrfunc) swig_varlink_setattr, /* tp_setattr */
       0,                                  /* tp_compare */
@@ -914,7 +918,11 @@ SwigPyObject_TypeOnce(void) {
       sizeof(SwigPyObject),                 /* tp_basicsize */
       0,                                    /* tp_itemsize */
       (destructor)SwigPyObject_dealloc,     /* tp_dealloc */
-      0,                                    /* tp_print */
+#if PY_VERSION_HEX < 0x030800b4
+      (printfunc)0,                         /*tp_print*/
+#else
+      (Py_ssize_t)0,                        /*tp_vectorcall_offset*/
+#endif
       (getattrfunc)0,                       /* tp_getattr */
       (setattrfunc)0,                       /* tp_setattr */
 #if PY_VERSION_HEX >= 0x03000000
@@ -1090,7 +1098,11 @@ SwigPyPacked_TypeOnce(void) {
       sizeof(SwigPyPacked),                 /* tp_basicsize */
       0,                                    /* tp_itemsize */
       (destructor)SwigPyPacked_dealloc,     /* tp_dealloc */
-      0,                                    /* tp_print */
+#if PY_VERSION_HEX < 0x030800b4
+      (printfunc)0,                         /*tp_print*/
+#else
+      (Py_ssize_t)0,                        /*tp_vectorcall_offset*/
+#endif
       (getattrfunc)0,                       /* tp_getattr */
       (setattrfunc)0,                       /* tp_setattr */
 #if PY_VERSION_HEX>=0x03000000

--- a/Source/Modules/python.cxx
+++ b/Source/Modules/python.cxx
@@ -4145,7 +4145,11 @@ public:
     printSlot(f, getSlot(n, "feature:python:tp_basicsize", tp_basicsize), "tp_basicsize");
     printSlot(f, getSlot(n, "feature:python:tp_itemsize"), "tp_itemsize");
     printSlot(f, getSlot(n, "feature:python:tp_dealloc", tp_dealloc_bad), "tp_dealloc", "destructor");
+    Printv(f, "#if PY_VERSION_HEX < 0x030800b4\n", NIL);
     printSlot(f, getSlot(n, "feature:python:tp_print"), "tp_print", "printfunc");
+    Printv(f, "#else\n", NIL);
+    printSlot(f, getSlot(n, "feature:python:tp_vectorcall_offset"), "tp_vectorcall_offset", "Py_ssize_t");
+    Printv(f, "#endif\n", NIL);
     printSlot(f, getSlot(n, "feature:python:tp_getattr"), "tp_getattr", "getattrfunc");
     printSlot(f, getSlot(n, "feature:python:tp_setattr"), "tp_setattr", "setattrfunc");
     Printv(f, "#if PY_VERSION_HEX >= 0x03000000\n", NIL);

